rfcs/0000-atomic-volatile.md

```diff
@@ -0,0 +1,669 @@
+Rust references. These unusual and unpleasant ergonomics are necessary in order
+to achieve the desired semantics of manually controlling every access to the
+target memory location, because the mere existence of a Rust reference pointing
+to a memory region allows the Rust compiler to generate memory operations
+targeting this region (be they prefetches, register spills...).
+
+Because a Rust pointer is not subjected to borrow checking and has no obligation
+of pointing towards a valid memory location, this means that using a Volatile
+wrapper in any way is unsafe.
+
+As a second difference, in addition to familiar atomic memory operations,
+volatile types expose the `load_not_atomic()` and `store_not_atomic()` methods.
+As their name suggest, these memory operations are not considered to be atomic
+by the compiler, and are therefore not safe to concurrently invoke in multiple
+threads.
+
+On hardware with global cache coherence, a category which encompasses the vast
+majority of Rust's supported compilation targets, use of these methods will
+generate exactly the same code as using the `load()` and `store()` atomic access
+methods with `Relaxed` memory ordering, with the only difference being that
+data races are Undefined Behavior from the compiler's point of view. When that
+is the case, safer `Relaxed` atomic volatile operations should be preferred to
+their non-atomic equivalents.
+
+Unfortunately, however, not all of Rust's compilation targets exhibit global
+cache coherence. GPU hardware, such as the `nvptx` target, may only exhibit
+cache coherence among local "blocks" of threads. And abstract machines like WASM
+may not guarantee cache coherence at all without specific precautions. On those
+compilation targets, `Relaxed` loads and stores may either be unavailable, or
+lead to the generation of machine instructions more complex than native loads
+and stores, which may not be wanted where maximal hardware control or CPU
+performance is desired.
+
+It is only for the sake of providing an alternative to the current
+`ptr::[read|write]_volatile` mechanism on such platforms that the
+`[load|store]_not_atomic()` functions are being proposed, and they should not
+be used where better alternative exists.
+
+Finally, unlike with atomics, the compiler is not allowed to optimize the above
+function into the following...
+
+```rust
+unsafe fn do_volatile_things(target: NonNull<VolatileU8>) -> u8 {
+    target.store(42, Ordering::Relaxed);
+    42
+}
+```
+
+...or even the following, which it could normally do if it the optimizer managed
+to prove that no other thread has access to the `VolatileU8` variable:
+
+```rust
+unsafe fn do_volatile_things(_target: NonNull<VolatileU8>) -> u8 {
+    42
+}
+```
+
+This is the definining characteristic of volatile operations, which makes them
+suitable for sensitive memory manipulations such as cryptographic secret erasure
+or memory-mapped I/O.
```

Comment:
Defining a language through the set of allowed optimizations doesn't work. We need a proper description of what the semantics are, from which the set of allowed optimizations can be *derived*.

I agree with having allowed and disallowed optimizations in the document as they help clarify intent, but they can never be "defining".

Reply:
I will try to rephrase this in terms of intent rather than definition, then.

Reply:
I tried to weaken the wording a bit in the latest revision to make absence of loads and store elision be one central _characteristic_ of volatile as opposed to being the thing that defines volatility, do you think that is enough?

Reply:
Sounds good, thanks!